core/src/apps/cardano/seed.py

```diff
@@ -30,30 +30,27 @@ def derive(self, node_path: list) -> bip32.HDNode:
         return node
 
 
-async def _get_passphrase(ctx: wire.Context) -> bytes:
-    passphrase = storage.cache.get_passphrase()
-    if passphrase is None:
-        passphrase = await protect_by_passphrase(ctx)
-        storage.cache.set_passphrase(passphrase)
-
-    return passphrase
-
-
 async def get_keychain(ctx: wire.Context) -> Keychain:
+    root = cache.get(cache.APP_CARDANO_ROOT)
+
     if not storage.is_initialized():
         raise wire.NotInitialized("Device is not initialized")
 
-    if mnemonic.is_bip39():
-        # derive the root node from mnemonic and passphrase
-        passphrase = await _get_passphrase(ctx)
-        root = bip32.from_mnemonic_cardano(mnemonic.get_secret().decode(), passphrase)
-    else:
-        seed = storage.cache.get_seed()
-        if seed is None:
-            passphrase = await _get_passphrase(ctx)
+    if root is None:
+        passphrase = await get_passphrase(ctx)
+        if mnemonic.is_bip39():
+            # derive the root node from mnemonic and passphrase
+            root = bip32.from_mnemonic_cardano(
+                mnemonic.get_secret().decode(), passphrase
+            )
+        else:
             seed = mnemonic.get_seed(passphrase)
-            storage.cache.set_seed(seed)
-        root = bip32.from_seed(seed, "ed25519 cardano seed")
+            root = bip32.from_seed(seed, "ed25519 cardano seed")
+
+        storage.cache.set(cache.APP_CARDANO_ROOT, root)
+
+    # let's not modify the one in the cache
+    root = root.clone()
```

Comment:
what's this about btw? couldn't we cache the derived root instead?

Reply:
Indeed we should. I am not sure why I have solved it this way, we should cache the derived(SEED_NAMESPACE) instead. Fixed in bc1fd06abab4ad92bde84249406326d38dfe937e